### PR TITLE
Bluetooth: controller: ISO resume ticker in ULL

### PR DESCRIPTION
**Pull request by mtpr-ot**
Added function ull_conn_iso_resume_ticker_start to enable LLL ISO
peripheral/central to start the one-shot resume ticker.
At timeout the common lll_resume handler is called, and based on the
LLL state, the peripheral/central is able to resume the CIG event.

The resume acts exactly like a normal event resume from the prepare
pipeline, with the exception that the LLL must provide specific
lll_event instances.

Signed-off-by: Morten Priess <mtpr@oticon.com>

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_conn_iso.c [hunk 1]
```diff
@@ -222,3 +229,97 @@ void ull_conn_iso_done(struct node_rx_event_done *done)
...
+
+	resume_event = param;
+	cig = resume_event->prepare_param.param;
+
+	/* Append timing parameters */
+	resume_event->prepare_param.ticks_at_expire = ticks_at_expire;
+	resume_event->prepare_param.remainder = remainder;
+	resume_event->prepare_param.lazy = 0;
+	resume_event->prepare_param.force = force;
+	resume_event->prepare_param.param = cig;
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:259]: It seems unnecessary to read `cig` out of `resume_event->prepare_param.param` and thereafter write it back into `resume_event->prepare_param.param`.

File: subsys/bluetooth/controller/ll_sw/ull_conn_iso.c [hunk 2]
```diff
@@ -222,3 +229,97 @@ void ull_conn_iso_done(struct node_rx_event_done *done)
...
+
+	/* Kick LLL resume */
+	ret = mayfly_enqueue(TICKER_USER_ID_ULL_HIGH, TICKER_USER_ID_LLL,
+			     0, &mfy);
+
+	LL_ASSERT(!ret);
+}
+
+void ull_conn_iso_resume_ticker_start(struct lll_event *resume_event,
+				      uint8_t  resume_cis,
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:270]: Would `resume_cis_handle` be a better name for this? Or maybe just `cis_handle`.

File: subsys/bluetooth/controller/ll_sw/ull_conn_iso.c [hunk 3]
```diff
@@ -222,3 +229,97 @@ void ull_conn_iso_done(struct node_rx_event_done *done)
...
+
+		cis = ll_conn_iso_stream_get(cig->cis_handles[resume_cis]);
+		acl = ll_conn_get(cis->lll.acl_handle);
+
+		ready_delay_us = lll_radio_rx_ready_delay_get(acl->lll.phy_rx, 1);
+#else
+	} else {
+		ready_delay_us = lll_radio_rx_ready_delay_get(0, 0);
+#endif /* CONFIG_BT_CTLR_PHY */
+	}
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:300]: Why use the if (0) construct? It is much more readable as:

```
#if defined(CONFIG_BT_CTLR_PHY)
￼	struct ll_conn_iso_stream *cis;
￼	struct ll_conn *acl;
￼
￼	cis = ll_conn_iso_stream_get(cig->cis_handles[resume_cis]);
￼	acl = ll_conn_get(cis->lll.acl_handle);
￼
￼	ready_delay_us = lll_radio_rx_ready_delay_get(acl->lll.phy_rx, 1);
#else
￼	ready_delay_us = lll_radio_rx_ready_delay_get(0, 0);
#endif /* CONFIG_BT_CTLR_PHY */
```

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_conn_iso.c (hunk 2 above)
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:270]: It is not a handle (this is zero based, handles are offset), so `_idx` or `_index` maybe?

File: subsys/bluetooth/controller/ll_sw/ull_conn_iso.c (hunk 3 above)
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:300]: Controller "boilerplate". @cvinayak, what is preferred?

File: subsys/bluetooth/controller/ll_sw/ull_conn_iso.c (hunk 2 above)
Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:270]: Right, I would say `_index`.

File: subsys/bluetooth/controller/ll_sw/ull_conn_iso.c (hunk 3 above)
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:300]: And also, there is a purpose of creating a scope for declaring local variables. With your suggeswtion, the variables will have to got to the top, with own #ifdef

Comment (thoh-ot) [reply subsys/bluetooth/controller/ll_sw/ull_conn_iso.c]: You can always create a scope, it's still more readable without the if/else

File: subsys/bluetooth/controller/ll_sw/ull_conn_iso.c [hunk 4]
```diff
@@ -226,3 +233,96 @@ void ull_conn_iso_done(struct node_rx_event_done *done)
...
+	static memq_link_t link;
+	static struct mayfly mfy = {0, 0, &link, NULL, lll_resume};
+	struct lll_event *resume_event;
+	struct lll_conn_iso_group *cig;
+	uint32_t ret;
+
+	LL_ASSERT(lazy == 0);
+
+	resume_event = param;
+	cig = resume_event->prepare_param.param;
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:256]: `cig` is unused

File: subsys/bluetooth/controller/ll_sw/ull_conn_iso.c (hunk 3 above)
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_conn_iso.c]: > You can always create a scope, it's still more readable without the if/else

That's true. I'll consider it and take input from other members.

**Review by cvinayak — CHANGES_REQUESTED** [review]
One change/clarification, otherwise looks good.

File: subsys/bluetooth/controller/ll_sw/lll.h [hunk 5]
```diff
@@ -113,6 +113,9 @@ enum {
 	TICKER_ID_CONN_ISO_BASE,
 	TICKER_ID_CONN_ISO_LAST = ((TICKER_ID_CONN_ISO_BASE) +
 				   (CONFIG_BT_CTLR_CONN_ISO_GROUPS) - 1),
+	TICKER_ID_CONN_ISO_RESUME_BASE,
+	TICKER_ID_CONN_ISO_RESUME_LAST = ((TICKER_ID_CONN_ISO_RESUME_BASE) +
+					  (CONFIG_BT_CTLR_CONN_ISO_GROUPS) - 1),
```
[the remark this answers is not in the document]
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ll_sw/lll.h:118]: > 1. As groups do not overlap, can one additional ticker be sufficient to be used as single shot? or

There is no guarantee that CIGs do not overlap, so having only a single ticker resource could lead to performance degradation.

[the remark this answers is not in the document]
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ll_sw/lll.h:118]: > 2\. Why not just use twice the value for `TICKER_ID_CONN_ISO_LAST` ?

This is for convenience. Without it you would have to calculate the ticker id offset. If you at some point would like to separate number of resume tickers from number of CIGs, then this also makes things easier.

[the remark this answers is not in the document]
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ll_sw/lll.h:118]: > 3\. Is this for CIS? if so, then a separate count for supported CIS per CIG should be used to allocate ticker id?

No, this is for CIGs, so only one resume ticker per CIG is needed.

**Review by cvinayak — APPROVED** [review]

**Review by wopu-ot — APPROVED** [review]

**Comment by cvinayak** [comment]
@thoh-ot Please re-review...

**Review by thoh-ot — APPROVED** [review]

**Review by kruithofa — APPROVED** [review]

**Pull request merged**